Clarify and organize cross browser testing strategies (#21668)
mdn/content: files/en-us/learn/tools_and_testing/cross_browser_testing/testing_strategies/index.md

  </tbody>
</table>

## Choosing which browsers and devices to test

Since you can't test every combination of browser and device, it's enough that you ensure your site works on the most important ones. In practical applications, "important" often means "commonly used among the target audience."

> You can classify browsers and devices by the amount of support you intend to give, e.g.:

Reviewer: ```suggestion
You can classify browsers and devices by the amount of support you intend to give. For example:
```

  </tbody>
</table>

## Choosing which browsers and devices to test

> Since it's impossible to test every combination of browser and device, it's enough that you ensure your site works on the most important ones. In practical applications, "important" often means "commonly used among the target audience."

Reviewer: ```suggestion
Since it's impossible to test every combination of browser and device, it's enough that you ensure your site works on the most important ones. In practical applications, "important" often means "commonly used among the target audience".
```

Reviewer 2: This is a regional difference--periods go inside quotes in American English, but outside quotes in British English. I don't know whether MDN is read by more British English users or American English users.

Reviewer: Periods don't go into the quotes if the quote is not a full sentence. I'm okay with (though I also won't write):

> He said, "I love you."

Reviewer: According to a whacky regex test over the corpus of MDN, we seem to use `".` much more commonly than `."` (>2000 times to >500 times). Still, it may be because more than half of my life's academic English training had been in Oxford English and it's just stuck with me subconsciously—I'm fine if you want to keep it this way. We can have an editorial consensus later since we have a lot of stylistic conventions to settle.

Reviewer 3: Overall on MDN we're not very prescriptive on fine style guide points like this. I would say in general, that if the [writing style guide](https://developer.mozilla.org/en-US/docs/MDN/Contribute/Guidelines/Writing_style_guide) isn’t explicit on a point, and if it’s not really obviously right or wrong one way or the other (to the point where it affects comprehension, I mean), then it doesn’t matter much, and it’s not really worthwhile to change docs from one to the other.

For our docs, clarity, helpfulness, and correctness are much more important than fine style guide points.

You can classify browsers and devices by the amount of support you intend to give, e.g.:

1. A grade: Common/modern browsers — Known to be capable. Test thoroughly and provide full support.
2. B grade: Older/less capable browsers — known not to be capable. Test, and provide a more basic experience that gives full access to core information and services.

> 3. C grade: Rare/unknown browsers — don't test, but assume they are capable. Serve the full site, which should work, at least with the fallbacks provided by our defensive coding.

Reviewer: You deleted the paragraph about defensive coding so this seems out of context. I suggest restoring that paragraph about graceful degradation.

Reviewer 2: Defensive coding is irrelevant to this paragraph, which is about choosing which browsers to test.

In the following sections, we'll build up a support chart in this format.

> **Note:** Yahoo first made this approach popular, with their [Graded browser Support](https://github.com/yui/yui3/wiki/Graded-Browser-Support) approach.

### Determine the most commonly-used browsers among your target audience

> Most people in North America or Western America use either Windows or Mac desktops or laptops, where the main browsers are Chrome, Firefox, Safari, IE, and Edge. You probably want to just test the latest versions of the first three, as these browsers receive regular updates. For Edge and IE, you probably want to test the last couple of versions; these should all go in the A grade tier.

Reviewer 3: ```suggestion
Most people in North America or Western Europe use either Windows or Mac desktops or laptops, where the main browsers are Chrome, Firefox, Safari, IE, and Edge. You probably want to just test the latest versions of the first three, as these browsers receive regular updates. For Edge and IE, you probably want to test the last couple of versions; these should all go in the A grade tier.
```

Reviewer: Edge is practically the same as Chrome, so it should be fine if you don't test it at all.

Reviewer 2: Agreed on all points; I just lumped Edge with IE to preserve the original grouping, and think it would be a good idea to change it, now that you mention it, given how much more modern Edge is than IE.

> Most people in North America or Western America use either Windows or Mac desktops or laptops, where the main browsers are Chrome, Firefox, Safari, IE, and Edge. You probably want to just test the latest versions of the first three, as these browsers receive regular updates. For Edge and IE, you probably want to test the last couple of versions; these should all go in the A grade tier.

Reviewer: This is far too America-centric for an opening paragraph—it's only supposed to be an example. If you find the original opening too informal, I suggest rewording it but at least keeping the same idea that we are "making educated guesses".

Reviewer 2: The bulk of MDN readers are based in either North America or Western Europe, and those who aren't are free to find information about their respective region. As you say, it's only supposed to be an example, not a comprehensive list of most-used browsers by region. 

I disagree that "making educated guesses" should be kept. Making educated guesses is merely a means to an end (the end being determining which browsers and devices one should test), and telling someone "make educated guesses" doesn't tell them anything new.

Reviewer 2: On second thought: an extra opening sentence seems redundant, given the section heading.

I'll preface the first sentence with "For example" and leave it at that. Flow is extremely important to me.

Many people use iOS and Android, so you probably also want to test the latest versions of iOS Safari, the last couple of versions of the old Android stock browser, and Chrome and Firefox for iOS and Android. You should ideally test these on both a phone and a tablet, to make sure that responsive designs are working OK.

Some people still use IE 9. This is old and less capable, so let's put it in the B grade tier.

Opera Mini isn't very capable in terms of running complex JavaScript at runtime. See [Opera Mini and JavaScript](https://dev.opera.com/articles/opera-mini-and-javascript/) for more details. We should put this into grade B as well.

> 

Reviewer 3: ```suggestion
```

(fix for markdownlint)

This gives us the following support chart so far:

1. A grade: Chrome and Firefox for Windows/Mac, Safari for Mac, Edge and IE for Windows (last two versions of each), iOS Safari for iPhone/iPad, Android stock browser (last two versions) on phone/tablet, Chrome and Firefox for Android (last two versions) on phone tablet
2. B grade: IE 9 for Windows, Opera Mini
3. C grade: n/a

If your target audience is mostly located somewhere else, then the most common browsers and OSs may differ from above.

> **Note:** "The CEO of my company uses a Blackberry, so we'd better make sure it looks good on that" can also be something to consider.

### Browser support stats

> Browser support stats can make your . For example:

Reviewer 2: Good point, will amend
